### PR TITLE
NEW API: GetObjectAttributes

### PR DESCRIPTION
**Pull request by zveinn**
This PR is related to https://github.com/minio/minio/pull/18732

P.s. the CI/CD tests will fail until we update the main repo with the new endpoint.

### CODE REVIEW

File: api-get-object-attributes.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/minio/minio-go/v7/pkg/encrypt"
+	"github.com/minio/minio-go/v7/pkg/s3utils"
+)
+
+// ObjectAttributesOptions ...
```
Comment (klauspost) [inline api-get-object-attributes.go:16]: ...

File: api-get-object-attributes.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		ChecksumSHA1   string `xml:",omitempty"`
+		ChecksumSHA256 string `xml:",omitempty"`
+	}
+	ObjectParts struct {
+		PartsCount           int
+		PartNumberMarker     int
+		NextPartNumberMarker int
+		MaxParts             int
+		IsTruncated          bool
+		Parts                []*struct {
```
Comment (klauspost) [inline api-get-object-attributes.go:73]:
```suggestion
		Parts                []ObjectPart `xml:"Part"`
```

File: api-get-object-attributes.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	"github.com/minio/minio-go/v7/pkg/s3utils"
+)
+
+// ObjectAttributesOptions ...
+// https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObjectAttributes.html
+type ObjectAttributesOptions struct {
+	ObjectAttributes     string
+	VersionID            string
+	MaxParts             int
+	PartNumberMarker     int
```
Comment (klauspost) [inline api-get-object-attributes.go:22]: I would kind of expect us to handle these internally and not expose these in the API.

So we just return all parts... I think this is overengineering on AWS' part already. We shouldn't make users bother with these and just return all parts.

File: api-get-object-attributes.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		ChecksumCRC32C string `xml:",omitempty"`
+		ChecksumSHA1   string `xml:",omitempty"`
+		ChecksumSHA256 string `xml:",omitempty"`
+	}
+	ObjectParts struct {
+		PartsCount           int
+		PartNumberMarker     int
+		NextPartNumberMarker int
+		MaxParts             int
+		IsTruncated          bool
```
Comment (klauspost) [inline api-get-object-attributes.go:72]: Again, all of these shouldn't really be exposed.

File: api-get-object-attributes.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+	"time"
+
+	"github.com/minio/minio-go/v7/pkg/encrypt"
+	"github.com/minio/minio-go/v7/pkg/s3utils"
+)
+
+// ObjectAttributesOptions ...
+// https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObjectAttributes.html
+type ObjectAttributesOptions struct {
+	ObjectAttributes     string
```
Comment (klauspost) [inline api-get-object-attributes.go:19]: It should be obvious how to use this SDK without needing to read the API specs. I don't really see any point in the user having to bother with filling this. 

We could accept a mask with named options - but I don't really see much point in filtering.

**Review by klauspost — COMMENTED** [review]
IMO you should make this easier to use and abstract away the overly complicated parts (pun not intended, but nice).

**Comment by zveinn** [comment]
> IMO you should make this easier to use and abstract away the overly complicated parts (pun not intended, but nice).

The api has been simplified based on our discussion today, I will double check AWS behavior tomorrow when it comes to large part numbers.

**Review by klauspost — APPROVED** [review]
Just unexport the const. LGTM otherwise.

File: api-get-object-attributes.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+// VersionID - The object version you want to attributes for
+// ServerSideEncryption - The server-side encryption algorithm used when storing this object in Minio
+type ObjectAttributesOptions struct {
+	VersionID            string
+	ServerSideEncryption encrypt.ServerSide
+}
+
+// ObjectAttributes ...
+type ObjectAttributes struct {
+	objectAttributesResponse
```
Comment (klauspost) [inline api-get-object-attributes.go:26]: Unexported. Must be available for caller.

File: api-get-object-attributes.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+	ObjectSize   int
+	Checksum     struct {
+		ChecksumCRC32  string `xml:",omitempty"`
+		ChecksumCRC32C string `xml:",omitempty"`
+		ChecksumSHA1   string `xml:",omitempty"`
+		ChecksumSHA256 string `xml:",omitempty"`
+	}
+	ObjectParts struct {
+		PartsCount int
+		Parts      []*objectPart `xml:"Part"`
```
Comment (klauspost) [inline api-get-object-attributes.go:60]: Type must be exported.

File: api-get-object-attributes.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+	if resp.StatusCode != http.StatusOK {
+		ER := new(ErrorResponse)
+		if err := xml.NewDecoder(resp.Body).Decode(ER); err != nil {
+			return ObjectAttributes{}, err
+		}
+
+		return ObjectAttributes{}, *ER
+	}
+
+	defer closeResponse(resp)
```
Comment (klauspost) [inline api-get-object-attributes.go:120]: Move up after error checking resp.

**Review by klauspost — CHANGES_REQUESTED** [review]

File: api-get-object-attributes.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+	urlValues.Add("attributes", "")
+	if opts.VersionID != "" {
+		urlValues.Add("versionId", opts.VersionID)
+	}
+
+	headers := make(http.Header)
+	headers.Set(amzObjectAttributes, GetObjectAttributesTags)
+
+	headers.Set(amzPartNumberMarker, "0")
+	headers.Set(amzMaxParts, "0")
```
[the remark this answers is not in the document]
Comment (zveinn) [reply api-get-object-attributes.go:94]: Made a comment about that above, will keep this unresolved until I test tomorrow

**Comment by klauspost** [comment]
@zveinn 

```
Running lint check
Error: api-get-object-attributes.go:65:1: exported: comment on exported type ObjectAttributePart should be of the form "ObjectAttributePart ..." (with optional leading article) (revive)
// ObjectAttributesResponse ...
^
Error: constants.go:63:2: exported: exported const GetObjectAttributesTags should have comment (or a comment on this block) or be unexported (revive)
	GetObjectAttributesTags = "ETag,Checksum,StorageClass,ObjectSize,ObjectParts"
	^
Error: make: *** [Makefile:15: lint] Error 1
```

**Comment by klauspost** [comment]
@zveinn 

Kinda interesting this gets thrown...

```
{"args":{"bucketName":"minio-go-test-rizxvdf5fd4dqwbk","objectName":"file2","opts":"minio.ObjectAttributesOptions{}"},"duration":6607,"error":"XML syntax error on line 1: invalid UTF-8","function":"GetObjectAttributes(ctx, bucketName, objectName, opts)","message":"GetObjectAttributes failed","name":"minio-go: testGetObjectAttributes","status":"FAIL"}
```

File: api-get-object-attributes.go [hunk 10]
```diff
@@ -0,0 +1,133 @@
+package minio
```
Comment (harshavardhana) [inline api-get-object-attributes.go:1]: Add missing copyright header.

File: api-get-object-attributes.go [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+	response := new(ObjectAttributesResponse)
+	if err := xml.NewDecoder(resp.Body).Decode(response); err != nil {
+		return err
+	}
+	o.ObjectAttributesResponse = *response
+
+	return
+}
+
+// ObjectAttributesResponse ...
```
Comment (klauspost) [inline api-get-object-attributes.go:49]: Fill out.

File: api-get-object-attributes.go [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+		ChecksumSHA1   string `xml:",omitempty"`
+		ChecksumSHA256 string `xml:",omitempty"`
+	}
+	ObjectParts struct {
+		PartsCount int
+		Parts      []*ObjectAttributePart `xml:"Part"`
+	}
+}
+
+// ObjectAttributePart ...
```
Comment (klauspost) [inline api-get-object-attributes.go:66]: Fill out...

File: api-get-object-attributes.go [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+type ObjectAttributePart struct {
+	ChecksumCRC32  string `xml:",omitempty"`
+	ChecksumCRC32C string `xml:",omitempty"`
+	ChecksumSHA1   string `xml:",omitempty"`
+	ChecksumSHA256 string `xml:",omitempty"`
+	PartNumber     int
+	Size           int
+}
+
+// GetObjectAttributes ...
```
Comment (klauspost) [inline api-get-object-attributes.go:76]: Make this proper documentation. Eg

```suggestion
// GetObjectAttributes provides detailed object information.
```

**Review by klauspost — COMMENTED** [review]
I guess for older versions it returns the object data?

Kinda clunky API-wise, but I guess AWS overloaded the `GET bucket/object` to implement this...

We should probably document that this is the behavior (if it is) when the API call isn't supported by the remote.. Maybe we should even check the response content type or something similar before decoding.

File: functional_tests.go [hunk 14]
```diff
@@ -12891,6 +13489,9 @@ func main() {
 
 	// execute tests
 	if isFullMode() {
+		testGetObjectAttributes()
+		testGetObjectAttributesErrorCases()
+		testGetObjectAttributesSSECEncryption()
```
Comment (harshavardhana) [inline functional_tests.go:13494]: move this inside `if tls` block @zveinn

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by harshavardhana** [comment]
Something is broken in the tests as well now because in this PR master branch runs fine with MinIO latest release.

```
{"args":{"bucketName":"minio-go-test-riqvc9tscw0inoj5","contentType":"","filePath":"","metadata":"custom/contenttype","objectName":"ugqepx3npjrsgnzcb1940qnoiikd5y","sse":{}},"duration":62,"function":"FPutEncryptedObject(bucketName, objectName, filePath, contentType, sse)","name":"minio-go: testSSES3EncryptionFPut","status":"PASS"}
{"args":{"bucketName":"minio-go-test-t1jmaabgxw9mey3i","objectName":"v33nnuhfn0gdkz1wbqjhmsx3gjn6gx"},"duration":4491,"function":"GetObject(bucketName, objectName)","name":"minio-go: testSSES3EncryptedGetObjectReadAtFunctional","status":"PASS"}
{"args":{"bucketName":"minio-go-test-3v21332559zpfsz4","objectName":"ob2ojnkoranpdpgvf5tsnnqch0vjgc"},"duration":7785,"function":"GetObject(bucketName, objectName)","name":"minio-go: testSSES3EncryptedGetObjectReadSeekFunctional","status":"PASS"}
{"args":{"destination":{"Bucket":"minio-go-test-rplpr2u4nzbis0ct","Object":"srcObject","Encryption":null,"UserMetadata":null,"ReplaceMetadata":false,"UserTags":null,"ReplaceTags":false,"LegalHold":"","Mode":"","RetainUntilDate":"0001-01-01T00:00:00Z","Size":0,"Progress":null},"source":{"Bucket":"minio-go-test-rplpr2u4nzbis0ct","Object":"srcObject","VersionID":"","MatchETag":"","NoMatchETag":"","MatchModifiedSince":"0001-01-01T00:00:00Z","MatchUnmodifiedSince":"0001-01-01T00:00:00Z","MatchRange":false,"Start":0,"End":0,"Encryption":{}}},"duration":262,"function":"CopyObject(destination, source)","name":"minio-go: testEncryptedSSES3ToSSES3CopyObject","status":"PASS"}
{"args":{"destination":{"Bucket":"minio-go-test-9q30ubdszad5fhmk","Object":"srcObject","Encryption":null,"UserMetadata":null,"ReplaceMetadata":false,"UserTags":null,"ReplaceTags":false,"LegalHold":"","Mode":"","RetainUntilDate":"0001-01-01T00:00:00Z","Size":0,"Progress":null},"source":{"Bucket":"minio-go-test-9q30ubdszad5fhmk","Object":"srcObject","VersionID":"","MatchETag":"","NoMatchETag":"","MatchModifiedSince":"0001-01-01T00:00:00Z","MatchUnmodifiedSince":"0001-01-01T00:00:00Z","MatchRange":false,"Start":0,"End":0,"Encryption":{}}},"duration":261,"function":"CopyObject(destination, source)","name":"minio-go: testEncryptedSSES3ToUnencryptedCopyObject","status":"PASS"}
{"args":{"destination":{"Bucket":"minio-go-test-1kleakxixnvtqvnj","Object":"dstObject","Encryption":{},"UserMetadata":null,"ReplaceMetadata":false,"UserTags":null,"ReplaceTags":false,"LegalHold":"","Mode":"","RetainUntilDate":"0001-01-01T00:00:00Z","Size":0,"Progress":null},"source":{"Bucket":"minio-go-test-1kleakxixnvtqvnj","Object":"srcObject","VersionID":"","MatchETag":"","NoMatchETag":"","MatchModifiedSince":"0001-01-01T00:00:00Z","MatchUnmodifiedSince":"0001-01-01T00:00:00Z","MatchRange":false,"Start":0,"End":0,"Encryption":null}},"duration":233,"function":"CopyObject(destination, source)","name":"minio-go: testUnencryptedToSSES3CopyObject","status":"PASS"}
{"args":{},"duration":151,"function":"CopyObjectPart(destination, source)","name":"minio-go: testUnencryptedToSSES3CopyObjectPart","status":"PASS"}
{"args":{},"duration":135,"function":"CopyObjectPart(destination, source)","name":"minio-go: testSSES3EncryptedToUnencryptedCopyPart","status":"PASS"}
{"args":{},"duration":169,"function":"CopyObjectPart(destination, source)","name":"minio-go: testSSES3EncryptedToSSES3CopyObjectPart","status":"PASS"}
```

However this PR is passing GetObjectAttributes but failing in other places.

**Review by klauspost — APPROVED** [review]
Lovely! <3

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**